File: members/U02BQ6G1SQJ/programmers/weeklychallenge/복서 정렬하기.py
def solution(weights, head2head):
    answerList = [];
    
    for r in range(len(head2head)):
        (winCnt, matchCnt, heavyCnt) = (0, 0, 0);
        
        for c in range(len(head2head)):
            if (head2head[r][c] == "W"):
                winCnt += 1;
                matchCnt += 1;
                
                if (weights[r] < weights[c]):
                    heavyCnt += 1;
                
            elif (head2head[r][c] == "L"):
                matchCnt += 1;
        
        winRate = (0 if (matchCnt == 0) else (winCnt / matchCnt));
        
        # print(winCnt, matchCnt, winRate, heavyCnt);
        
        answerList.append([r + 1, winRate, heavyCnt, weights[r]]);
    
    # print(answerList);
    
    return [k[0] for k in sorted(answerList, key = lambda k : (-k[1], -k[2], -k[3], k[0]))];

File: members/U02BWCHJA6Q/Programers/weekly_challenge/복서_정렬하기.js
// https://programmers.co.kr/learn/courses/30/lessons/85002

function solution(weights, head2head) {
  return head2head
    .map((match, index) => {
      let total = 0;
      let win = 0;
      let winByHeavy = 0;
      const weight = weights[index];

      for (let i = 0; i < match.length; i++) {
        if (match[i] === 'N') {
          continue;
        }

        total++;
        if (match[i] === 'W') {
          win++;

          if (weights[i] > weight) {
            winByHeavy++;
          }
        }
      }

      const winRate = win === 0 ? 0 : win / total;

      return {
        winRate,
        winByHeavy,
        weight,
        index: index + 1,
        weight: weights[index],
      };
    })
    .sort((a, b) => {
      if (a.winRate !== b.winRate) {
        return b.winRate - a.winRate;
      }
      if (a.winByHeavy !== b.winByHeavy) {
        return b.winByHeavy - a.winByHeavy;
      }
      if (a.weight !== b.weight) {
        return b.weight - a.weight;
      }
      return a.index - b.index;
    })
    .map(({ index }) => index);
}
